tea.py

```diff
@@ -0,0 +1,460 @@
+
+    def __init__(self, parameter_values=None, results=None):
+        self.parameter_values = parameter_values
+        if results is not None:
+            self.results = results
+        else:
+            self.results = {}
+
+        if parameter_values is None and results is None:
+            print(
+                "Error: Missing parameter_values and results, please supply parameters to TEA()"
+            )
+        if parameter_values is None and results is not None:
+            print(
+                "Warning: Missing parameter_values, values from results are used."
+            )
+        
+        super().__init__()
+
+    def stack_energy_densities(self, inputs=None, print_values=False):
+        # Inputs: "Electrolyte density [kg.m-3]"
+        if inputs is None:
+            # print("Warning: Missing input parameters")
+            inputs = {}
+        else:
+            inputs = inputs
+
+        # ocv's
+        # add optional user defined cut-off voltages (supplied with inputs)?
+        ne_ocv = self.parameter_values["Negative electrode OCP [V]"]
+        pe_ocv = self.parameter_values["Positive electrode OCP [V]"]
+        x0, x100, y100, y0 = pybamm.lithium_ion.get_min_max_stoichiometries(
+            self.parameter_values
+        )
+        soc = pybamm.linspace(0, 1)
+        x = x0 + soc * (x100 - x0)
+        y = y0 - soc * (y0 - y100)
+
+        self.results["Negative electrode average OCP [V]"] = self.results.get("Negative electrode average OCP [V]") or ne_ocv(x).entries.mean()
+        self.results["Positive electrode average OCP [V]"] = self.results.get("Positive electrode average OCP [V]") or pe_ocv(y).entries.mean()
+        self.results["Stack average OCP [V]"] = self.results.get("Stack average OCP [V]") or self.results.get("Positive electrode average OCP [V]") - self.results.get("Negative electrode average OCP [V]")
+
+        # areal capacity
+        # add optional user defined 1st cycle capacity loss in case capacity is supplied explicitly (with results) or on top of "Formation capacity loss [mA.h.cm-2]"?
+        # update capacity, in case of optional user defined cut-off voltages (supplied with inputs)?
+        param = pybamm.LithiumIonParameters()
+        esoh_solver = pybamm.lithium_ion.ElectrodeSOHSolver(
+            self.parameter_values, param
+        )
+        Q_n = self.parameter_values.evaluate(param.n.Q_init)
+        Q_p = self.parameter_values.evaluate(param.p.Q_init)
+        Q_Li = self.parameter_values.evaluate(param.Q_Li_particles_init)
+        inputs_ = {"Q_Li": Q_Li, "Q_n": Q_n, "Q_p": Q_p}
+        sol = esoh_solver.solve(inputs_)
+        areal_capacity = sol[
+            "Capacity [mA.h.cm-2]" # or better min("sol[Q_n * (x_100 - x_0)]"", "sol[Q_p * (y_0 - y_100)]")?
+        ]
+        self.results["Capacity [mA.h.cm-2]"] = self.results.get("Capacity [mA.h.cm-2]") or areal_capacity
+
+        # thicknesses
```

Comment:
the logic for all these calculations is very confusing. I think you should do some calculations of things that are fixed in the `__init__`. E.g. a have dict (maybe called `derived_values` or something) to store the results you need from the SOH solution, store the average OCVs). Then in the functions just do the calculations using values from `self.parameter_values` and `self.derived_values`. I'm not sure why `results` needs to store things that are already in `parameter_value` (e.g. thicknesses).